core-bundle/contao/modules/ModuleLogin.php

```diff
@@ -105,20 +112,14 @@ protected function compile()
 
 		$container = System::getContainer();
 		$request = $container->get('request_stack')->getCurrentRequest();
+		$security = $container->get('security.helper');
+		$user = $security->getUser();
 		$exception = null;
 		$lastUsername = '';
+		$isRemembered = $security->isGranted('IS_REMEMBERED');
+		$isTwoFactorInProgress = $security->isGranted('IS_AUTHENTICATED_2FA_IN_PROGRESS');
 
-		// Only call the authentication utils if there is an active session to prevent starting an empty session
-		if ($request?->hasSession() && ($request->hasPreviousSession() || $request->getSession()->isStarted()))
-		{
-			$authUtils = $container->get('security.authentication_utils');
-			$exception = $authUtils->getLastAuthenticationError();
-			$lastUsername = $authUtils->getLastUsername();
-		}
-
-		$authorizationChecker = $container->get('security.authorization_checker');
-
-		if ($authorizationChecker->isGranted('ROLE_MEMBER'))
+		if ($request && $user instanceof FrontendUser && !$isTwoFactorInProgress && (!$isRemembered || $objPage->type != 'error_401'))
```

Comment:
Why did you add the `$objPage->type != 'error_401'` check here?

Reply:
Also, the `$request` check seems unrelated.

Show the "you are logged in as …" message if

* the user is a `FrontendUser`
* and we are not in the middle of a 2FA process
* and the user is fully authenticated **or the page is not a 401 page**.

This means that the same login form behaves differently on a regular page and on a 401 page. If an installation uses the 401 page to redirect to a regular page with a login form, the reauthentication options are never shown.